### DISCUSSION TITLE
DFG LUT and roughness doubt (perceptual roughness vs α)

### DISCUSSION THREAD

**Discussion by TaaTT4** [google/filament#5421] [Q&A] [answered]
Looking at cmgen code I've seen that the y coord is [re-mapped](https://github.com/google/filament/blob/main/libs/ibl/src/CubemapIBL.cpp#L1024) as `linear_roughness = coord_y * coord_y`.  This `linear_roughness` is then provided as input (together with `NoV = coord_x`) to the DFG LUT generator algorithm. Correct me if I'm wrong, this corresponds to say that `perceptualRoughness = coord_y` and `α = linear_roughness`. This indeed is consistent with what is stated in the [documentation](https://google.github.io/filament/Filament.md.html#toc5.3.4.3) about how to calculate the two DFG terms (where their values depend solely on `n⋅v` and `α`).
But shouldn't the same parameters being used also when the DFG LUT texture is [sampled](https://github.com/google/filament/blob/main/shaders/src/shading_lit.fs#L222)? I could be wrong, but since `perceptualRoughness = PixelParams.perceptualRoughness` and `α = PixelParams.roughness` this leads in using two different mappings for the DFG LUT creation and the data retrieval. As the code is now, it's like saying `dfg_lut_texture = DFG(n⋅v, α)` (in texture creation) and `dfg_lut_value = DFG(n⋅v, perceptualRoughness)` (in texture sampling) which sounds wrong to me for what I've understood so far about PBR.

**Comment by romainguy** [comment, marked as answer]
There's a comment to that effect in the code:
`// PrefilteredDFG_LUT() takes a LOD, which is sqrt(roughness) = perceptualRoughness`

What we are actually doing is using `sqrt(linear_roughness)` (to undo the encoding of `linear_roughness = coord.y ** 2`) or `sqrt(coord_y * coord_y)` which simplifies down to `coord_y` or `perceptualRoughness` in this case.

**Reply by TaaTT4** [reply]
I'm stupid. I've been mislead by the fact that the roughness remapping (from perceptual to linear) happens outside the DFV function. Thanks for the answer!
